### PR TITLE
[stable/mongodb] - Refactor MongoDB chart to support replica set configurations

### PR DESCRIPTION
**Pull request by tompizmor**
**What this PR does / why we need it**:
This PR adds support to the stable/mongodb chart to be deployed as a replica set configuring primary, secondary and arbiter nodes.

Other changes:
- When deploying in replica-set mode, it uses stateful-sets with headless-svc and volumeClaimTemplates for persistence.
- When using replica-set mode, pod-disruption-budgets are configurable too 
- It allows to overwrite the default config file with a config map
- replica set parameters are configurable through values.yaml
- Two `values.yaml` availables, `values.yaml` and `values-production.yaml`

**Special notes for your reviewer**:
Please **do not merge** until I say the contrary as I need to make publish the new version of the docker image.

### CODE REVIEW

File: stable/mongodb/values.yaml
Comment (juan131) [inline stable/mongodb/values.yaml]: Why did you change this parameter name? By the way, it's not aligned with the `README.md`

File: stable/mongodb/values.yaml
Comment (juan131) [inline stable/mongodb/values.yaml]: `imageTag` -> `image.tag`

File: stable/mongodb/values.yaml
Comment (juan131) [inline stable/mongodb/values.yaml:46]: Why did you move this line?

File: stable/mongodb/values-production.yaml
Comment (juan131) [inline stable/mongodb/values-production.yaml]: Why did you change this parameter name? By the way, it's not aligned with the README.md

File: stable/mongodb/values-production.yaml
Comment (juan131) [inline stable/mongodb/values-production.yaml]: `imageTag` -> `image.tag`

File: stable/mongodb/values-production.yaml
Comment (juan131) [inline stable/mongodb/values-production.yaml:50]: Why did you move this line?

File: stable/mongodb/templates/deployment-standalone.yaml
Comment (juan131) [inline stable/mongodb/templates/deployment-standalone.yaml]: Define a function on *_helpers.tpl* for this and use it.

File: stable/mongodb/templates/deployment-standalone.yaml
Comment (juan131) [inline stable/mongodb/templates/deployment-standalone.yaml]: Please remove this

File: stable/mongodb/templates/headless-svc-rs.yaml
Comment (juan131) [inline stable/mongodb/templates/headless-svc-rs.yaml:12]: Should this be configurable?

**Review by juan131 — CHANGES_REQUESTED** [review]

File: stable/mongodb/templates/headless-svc-rs.yaml
Comment (tompizmor) [reply stable/mongodb/templates/headless-svc-rs.yaml:12]: nop, this one should be ClusterIp and ClusterIp: Nono

File: stable/mongodb/templates/deployment-standalone.yaml
Comment (tompizmor) [reply stable/mongodb/templates/deployment-standalone.yaml]: ok

File: stable/mongodb/values.yaml
Comment (tompizmor) [reply stable/mongodb/values.yaml]: human mistake!

File: stable/mongodb/templates/deployment-standalone.yaml
Comment (tompizmor) [reply stable/mongodb/templates/deployment-standalone.yaml]: ok

File: stable/mongodb/values-production.yaml
Comment (tompizmor) [reply stable/mongodb/values-production.yaml:50]: I see it fine

**Comment by unguiculus** [comment]
Why do we need this? There's already https://github.com/kubernetes/charts/tree/master/stable/mongodb-replicaset.

cc @prydonius

**Comment by tompizmor** [comment]
@unguiculus We have done similar PRs for Rabbitmq (https://github.com/kubernetes/charts/pull/4591) and Redis (https://github.com/kubernetes/charts/pull/4662) too.

The configuration for mongodb and mongodb-replicaset are a bit different. 

* mongodb-replicaset deploys X mongodb  replicas and one of the is the master (you need to find which one). The mongodb chart has a fixed stateful set for the master. 
* mongodb-replicaset chart does not use arbiter nodes while mongodb chart does
* The replication is disabled by default in the mongodb chart, but it can be easily enabled, so it is easy to move applications from a single mongodb node to a replicaset scenario.

cc/ @prydonius @sameersbn what do you think?

File: stable/mongodb/README.md
Comment (sameersbn) [inline stable/mongodb/README.md]: The default value column should just say `Always`

File: stable/mongodb/templates/statefulset-arbiter-rs.yaml
Comment (sameersbn) [inline stable/mongodb/templates/statefulset-arbiter-rs.yaml:3]: the arbiter nodes do not save any data, do we really need to use a statefulset here?

File: stable/mongodb/values-production.yaml
Comment (sameersbn) [inline stable/mongodb/values-production.yaml]: should we move `port` to `service.port`?

File: stable/mongodb/values-production.yaml
Comment (sameersbn) [inline stable/mongodb/values-production.yaml]: remove this comment

File: stable/mongodb/values-production.yaml
Comment (sameersbn) [inline stable/mongodb/values-production.yaml:136]: we should be able to completely disable the probes

File: stable/mongodb/README.md
Comment (sameersbn) [inline stable/mongodb/README.md]: we should remove this line. It does not make sense anymore.
We need to add a section on deploying the chart with values-production.yaml configuration (see https://github.com/kubernetes/charts/tree/master/stable/wordpress#production-and-horizontal-scaling)

**Review by sameersbn — CHANGES_REQUESTED** [review]

File: stable/mongodb/README.md
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/templates/statefulset-arbiter-rs.yaml
Comment (tompizmor) [reply stable/mongodb/templates/statefulset-arbiter-rs.yaml:3]: we needed so the network identifier is stable.

File: stable/mongodb/values-production.yaml
Comment (tompizmor) [reply stable/mongodb/values-production.yaml]: done

File: stable/mongodb/values-production.yaml
Comment (tompizmor) [reply stable/mongodb/values-production.yaml]: done

File: stable/mongodb/values-production.yaml
Comment (tompizmor) [reply stable/mongodb/values-production.yaml:136]: done

File: stable/mongodb/README.md
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md
Comment (sameersbn) [inline stable/mongodb/README.md]: I would change this to,

The [values-production.yaml](values-production.yaml) file consists a configuration to deploy a scalable MongoDB deployment for production environments. We recommend that you base your production configuration on this template and adjust the parameters appropriately.

```bash
$ curl -O https://raw.githubusercontent.com/kubernetes/charts/master/stable/mongodb/production-values.yaml
$ helm install --name my-release -f values-production.yaml stable/mongodb
```

File: stable/mongodb/README.md
Comment (sameersbn) [inline stable/mongodb/README.md]: `stateful set` -> `statefulset`

File: stable/mongodb/README.md
Comment (sameersbn) [inline stable/mongodb/README.md]: I think name of the statefulset in the command is incorrect. Should probably be 

```bash
$ kubectl scale statefulset my-release-mongodb-secondary --replicas=3 
```

Ideally we should be documenting the `helm upgrade` way but I understand that the use of random passwords is the issue here

**Review by sameersbn — CHANGES_REQUESTED** [review]

File: stable/mongodb/README.md
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md
Comment (tompizmor) [reply stable/mongodb/README.md]: fixed

**Review by sameersbn — APPROVED** [review]

**Comment by sameersbn** [comment]
/approve /ok-to-test

Please wait for approval by @unguiculus before the merge.

**Comment by tompizmor** [comment]
ping @unguiculus

**Comment by unguiculus** [comment]
I'm not going to block this but was hoping @prydonius would jump in. We had a similar discussion on another chart. I think it was Rabbitmq.

**Comment by prydonius** [comment]
I think this is okay, but we should outline the differences you mentioned in https://github.com/kubernetes/charts/pull/5578#issuecomment-392482527 in the Chart's README in a section about enabling high-availability.

**Comment by tompizmor** [comment]
@prydonius, updated!

File: stable/mongodb/README.md
Comment (juan131) [inline stable/mongodb/README.md]: Why has this being removed?

Comment (tompizmor) [reply stable/mongodb/README.md]: mistake, fixed

**Review by juan131 — APPROVED** [review]
Thanks!

File: stable/mongodb/README.md
Comment (sabbour) [inline stable/mongodb/README.md:115]: @tompizmor in the README, the command to install the Helm chart with replication is incorrect. `replication.enabled=true` should be `replicaSet.enabled=true`.

**Pull request merged**
